### PR TITLE
[WPF] Input Evolution

### PR DESCRIPTION
**Pull request by almedina-ms**
## Related Issue
Part of #3081 

## Description
Adds support for Labels and Input validation on the WPF renderer. This includes changes to the InputBindings behaviour that has to be documented for users of v1.3.

## How Verified
To be added

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/4172)

### CODE REVIEW

File: source/dotnet/Library/AdaptiveCards/Rendering/ErrorMessageConfig.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace AdaptiveCards.Rendering
+{
+    /// <summary>
+    /// Properties which control rendering of media
```
Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards/Rendering/ErrorMessageConfig.cs:9]: >rendering of media [](start = 33, length = 18)

nit: outdated comments

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@ public static FrameworkElement Render(AdaptiveActionSet actionSet, AdaptiveRende
             return outerActionSet;
         }
 
-        public static void AddRenderedActions(Grid uiContainer, IList<AdaptiveAction> actions, AdaptiveRenderContext context, AdaptiveInternalID actionSetId)
+        public static void AddRenderedActions(Grid uiContainer, IList<AdaptiveAction> actions, AdaptiveRenderContext context,  AdaptiveInternalID actionSetId)
```
Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs:32]: >  [](start = 125, length = 1)

nit: extra spacing

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs [hunk 3]
```diff
@@ -89,8 +89,15 @@ public static void AddRenderedActions(Grid uiContainer, IList<AdaptiveAction> ac
 
                 foreach (AdaptiveAction action in actionsToProcess)
                 {
+                    var rendereableAction = context.GetRendereableElement(action);
+
+                    if (rendereableAction is AdaptiveSubmitAction)
+                    {
+                        context.SubmitActionCardId[rendereableAction as AdaptiveSubmitAction] = context.RenderArgs.ContainerCardId;
```
Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs:96]: nit: is there a reason why SubmitActionCardId has AdaptiveSubmitAction as a key?

Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs:96]: does it create a new internal id?

Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs:96]: >ContainerCardId [](start = 115, length = 15)

Will it create a new adaptive internal id?

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRenderContext.cs [hunk 4]
```diff
@@ -465,13 +472,96 @@ public void ToggleShowCardVisibility(Button uiAction)
...
+            InputBindings.Clear();
+
+            List<string> inputsToValidate = RetrieveInputList(submitAction);
+
+            // Iterate through all the elements and validate them
+            foreach (string inputId in inputsToValidate)
+            {
+                AdaptiveInputValue inputValue = InputValues[inputId];
+                bool inputIsValid = inputValue.Validate();
+                allInputsValid = allInputsValid && inputIsValid;
```
Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRenderContext.cs:499]: could be written as allInputsValid  &= inputIsValid;

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs (hunk 3 above)
Comment (almedina-ms) [reply source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs:96]: I had to use this as it's the only data we have when the action is invoked

---
In reply to: [440532423](https://github.com/microsoft/AdaptiveCards/pull/4172#discussion_r440532423) [](ancestors = 440532423)

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRenderContext.cs [hunk 5]
```diff
@@ -465,13 +472,96 @@ public void ToggleShowCardVisibility(Button uiAction)
...
+
+                // If the validation failed, set focus to the first element that failed
+                if (!allInputsValid && !firstInvalidInputFound)
```
Comment (jwoo-msft) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRenderContext.cs:502]: >firstInvalidInputFound [](start = 40, length = 22)

nit: this is perfectly good. it was less clear to me compared to having inputIsValid in the if statement. ex. if allinputs are not valid and first Invalid Input has not been found vs. if input is not valid and first invalid input has

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedTimeInput.cs [hunk 6]
```diff
@@ -7,6 +7,7 @@
 
 namespace AdaptiveCards.Rendering.Wpf
 {
+
     public static class XceedTimeInput
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedTimeInput.cs:11]: Nit: Stray newline

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedTimeInput.cs [hunk 7]
```diff
@@ -32,14 +33,16 @@ public static FrameworkElement Render(AdaptiveTimeInput input, AdaptiveRenderCon
                     timePicker.Value = value;
                 TimeSpan minValue;
                 if (IsSupportedTimeFormat(input.Min) && TimeSpan.TryParse(input.Min, out minValue))
-                    timePicker.EndTime = minValue;
+                    timePicker.StartTime = minValue;
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedTimeInput.cs:36]: can you explain this change a bit on removing the default EndTime initialization here?

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedTimeInput.cs [hunk 8]
```diff
@@ -32,14 +33,16 @@ public static FrameworkElement Render(AdaptiveTimeInput input, AdaptiveRenderCon
...
                 TimeSpan maxValue;
                 if (IsSupportedTimeFormat(input.Max) && TimeSpan.TryParse(input.Max, out maxValue))
                     timePicker.EndTime = maxValue;
                 timePicker.Watermark = input.Placeholder;
                 timePicker.Style = context.GetStyle("Adaptive.Input.Time");
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedTimeInput.cs:41]: >"Adaptive.Input.Time") [](start = 52, length = 22)

nitnit: If you are in here and fixing it, can you add in opening and closing braces even for single line statements? Thanks!

**Comment by shalinijoshi19** [comment]
>             return (dateString != null) && textFunctionRegex.IsMatch(dateString);

For efficiency, we should change this to dateString.IsNullOrWhiteSpace() to catch whitespaces upfront before calling the regex operation.

---
Refers to: source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedDateInput.cs:22 in 506751c. [](commit_id = 506751c7d33af8ba8bc9018c89e33dec0971c0ba, deletion_comment = False)

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedInputValue.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+using System.Windows;
+using System.Windows.Controls;
+using System.Text.RegularExpressions;
+using Xceed.Wpf.Toolkit;
+
+namespace AdaptiveCards.Rendering.Wpf
+{
+
+    public class AdaptiveXceedTextInputValue : AdaptiveTextInputValue
+    {
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedInputValue.cs:13]: Nit: Additional newline creep line 11

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedInputValue.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        public AdaptiveXceedTimeInputValue(AdaptiveTimeInput inputElement, UIElement renderedElement) : base(inputElement, renderedElement) { }
+
+        public override string GetValue()
+        {
+            return ToIso8601Time((RenderedInputElement as TimePicker).Text);
+        }
+    }
+
+}
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf.Xceed/XceedInputValue.cs:52]: Nit: Additional newline creep line 51

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ControlExtensions.cs [hunk 11]
```diff
@@ -48,6 +48,14 @@ public static void SetColor(this Span inlineRun, AdaptiveTextColor color, bool i
...
         }
 
+        public static void SetColor(this Inline inlineRun, AdaptiveTextColor color, bool isSubtle, AdaptiveRenderContext context)
+        {
+            FontColorConfig colorOption = context.GetForegroundColors(color);
+            string colorCode = isSubtle ? colorOption.Subtle : colorOption.Default;
+            inlineRun.Foreground = context.GetColorBrush(colorCode);
+        }
+
+
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ControlExtensions.cs:58]: nit: additional/stray newline creep line 58

File: source/dotnet/Library/AdaptiveCards/Rendering/InputsLabelConfig.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace AdaptiveCards.Rendering
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards/Rendering/InputsLabelConfig.cs:6]: Naming: InputLabelConfig and InputLabelsConfig as referring to the label configuration and label collection configuration seems pretty confusing.  What's the point of having these split into 2 separate classes again? let me know if you discuss this elsewhere in the spec wherei should be looking..

**Comment by shalinijoshi19** [comment]
@almedina-ms can we get this readied for checking in? (I think the .NET CI was failing still)

**Pull request merged**